src/common/visualization/visualization.py

```diff
@@ -0,0 +1,81 @@
+# SPDX-FileCopyrightText: 2023 spdx contributors
+#
+# SPDX-License-Identifier: Apache-2.0
+from typing import Dict, List
+
+try:
+    from networkx import DiGraph
+except ImportError:
+    DiGraph = None
+from spdx.document_utils import get_contained_spdx_element_ids, get_element_from_spdx_id
+from spdx.model.document import Document
+from spdx.model.relationship import Relationship
+
+
+def visualize_document(document: Document, visualize: str) -> None:
+    try:
+        from networkx.drawing import nx_agraph
+    except ImportError:
+        raise ImportError
+
+    graph = generate_graph_from_spdx(document)
+    _color_nodes(graph)
+    attributes_graph = nx_agraph.to_agraph(graph)  # convert to a graphviz graph
+    attributes_graph.draw(visualize, prog="dot")
+
+
+def generate_graph_from_spdx(document: Document) -> DiGraph:
+    try:
+        from networkx import DiGraph
+    except ImportError:
+        raise ImportError
+    graph = DiGraph()
+    graph.add_node(document.creation_info.spdx_id, element=document.creation_info)
+
+    contained_elements: List[str] = get_contained_spdx_element_ids(document)
+    contained_element_nodes = [
+        (spdx_id, {"element": get_element_from_spdx_id(document, spdx_id)}) for spdx_id in contained_elements
+    ]
+    graph.add_nodes_from(contained_element_nodes)
+
+    relationships_by_spdx_id: Dict[str, List[Relationship]] = _get_relationships_by_spdx_id(document.relationships)
+
+    for spdx_id, relationships in relationships_by_spdx_id.items():
+        if spdx_id not in graph.nodes():
+            graph.add_node(spdx_id, element=get_element_from_spdx_id(document, spdx_id))
```

Comment:
Didn't you already add all contained elements in line 39? So `get_element_from_spdx_id(document, spdx_id)` always yields `None` here, doesn't it?

Reply:
Yes, this line will add any external spdx_id if present so `get_element_from_spdx_id` yields `None`. For the later use of the graph, it would be helpful if the node has some kind of flag indicating that the node represents an element (including external elements), so I would keep the node attribute `element` and fill it with `None`.